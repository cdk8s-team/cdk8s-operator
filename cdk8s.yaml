operator: node test/fixtures/echo-app.js

