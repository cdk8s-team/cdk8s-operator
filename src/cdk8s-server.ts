// cdk8s-server: http server which synthesizes cdk operators
// ------------------------------------------------------------------------------------
// request body should include an input manifest for a single resource
// response will include the synthesized output manifest

import * as http from 'http';
import { spawn } from 'child_process';
import * as yaml from 'yaml';
import * as fs from 'fs';
import * as stream from 'stream';

const config = yaml.parse(fs.readFileSync('cdk8s.yaml', 'utf-8'));

const command = config.app;
if (!command) {
  throw new Error('cdk8s.yaml is missing an "app" attribute');
}

// needed in docker
process.on('SIGINT', function() {
  process.exit();
});

const server = http.createServer((req, res) => {

  const input = fs.createWriteStream('/tmp/input.json');
  req.pipe(input);
  input.on('close', () => {
    const child = spawn(`${command} /tmp/input.json`, {
      stdio: [ 'pipe', 'pipe', 'pipe' ],
      shell: true,
    });

    const stderr = new Array<Buffer>();

    child.stdout.on('data', chunk => res.write(chunk));
    child.stderr.on('data', chunk => stderr.push(chunk));

    const flusherr = (out: stream.Writable) => {
      for (const c of stderr) {
        out.write(c);
      }
    };

    child.on('exit', code => {
      if (code === 0) {
        flusherr(process.stderr); // flush stderr to server terminal
        return res.end();
      }

      res.statusCode = 400; // bad request

      flusherr(res);

      res.end();
    });

  });

});

server.listen(8080);

console.error('cdk8s-server listening on 8080');
console.error('- Request body should include a single k8s resource in json');
console.error(`- Request will be piped through STDIN to ${command}`)
console.error('- Response is the STDOUT and expected to be a multi-resource yaml manifest');

