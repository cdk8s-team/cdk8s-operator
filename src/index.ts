export * from './operator';
export * from './server';