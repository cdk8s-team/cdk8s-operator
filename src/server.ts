import { spawn } from 'child_process';
import { createWriteStream, mkdtempSync, readFileSync } from 'fs';
import { createServer, IncomingMessage, Server as HttpServer, ServerResponse } from 'http';
import { tmpdir } from 'os';
import { join } from 'path';

export interface ServerProps {
  /**
   * The command to execute in order to synthesize the CDK app.
   */
  readonly appCommand: string;
}

export class Server {
  private readonly server: HttpServer;
  private readonly appCommand: string;
  private readonly tmpdir: string;

  constructor(props: ServerProps) {
    this.appCommand = props.appCommand;

    this.server = createServer((req, res) => this.handleRequest(req, res).catch(e => {
      console.error('server error: ', e);
      res.statusCode = 500;
      res.write(e.message);
      res.end();
    }));

    this.tmpdir = mkdtempSync(join(tmpdir(), 'cdk8s-operator-'));
  }

  /**
   * Starts HTTP server.
   * @param port The port to listen to. If not specified, the `PORT` environment
   * variable will be used. If that's not specified an available port will be
   * auto-selected.
   */
  public async listen(port?: number): Promise<number> {
    const lport = port ?? process.env.PORT ?? 0;
    return new Promise((ok, ko) => {
      this.server.listen(lport, () => {
        const addr = this.server.address();
        if (typeof(addr) === 'string') {
          throw new Error(`cannot determine port from server address ${addr}`);
        }

        return ok(addr.port);
      });

      this.server.on('error', err => ko(err));
    });
  }

  /**
   * Stop server.
   */
  public close() {
    this.server.close();
  }

  private async handleRequest(req: IncomingMessage, res: ServerResponse) {
    const inputfile = await this.writeInputFile(req);

    const child = spawn(this.appCommand, [inputfile], {
      stdio: ['ignore', 'pipe', 'pipe'],
      shell: true,
    });

    const stderr = new Array<Buffer>();

    res.setHeader('Content-Type', 'application/json');

    // stdout should go directly to the response
    child.stdout.on('data', chunk => {
      process.stderr.write('output: ' + chunk);
      res.write(chunk);
    });

    // for stderr: write to server terminal and only send back if we exited with a non-zero
    child.stderr.on('data', chunk => {
      process.stderr.write(chunk);
      stderr.push(chunk);
    });

    // will be caused by the async handler and 500 will be returned.
    child.on('error', err => {
      throw err;
    });

    child.on('exit', code => {
      if (code !== 0) {
        res.statusCode = 500;
        for (const c of stderr) {
          res.write(c);
        }
        res.end();
      }

      // success
      return res.end();
    });
  }

  private async writeInputFile(req: IncomingMessage): Promise<string> {
    return new Promise((ok, ko) => {
      const inputfile = join(this.tmpdir, `input-${Math.round(Math.random() * 999999)}.json`);
      const input = createWriteStream(inputfile);
      req.pipe(input);

      input.on('close', () => {
        try {
          const inputJson = JSON.parse(readFileSync(inputfile, 'utf-8'));
          console.error(`input: ${JSON.stringify(inputJson)}`);
          return ok(inputfile);
        } catch (e) {
          return ko(new Error(`unable to parse request body as JSON: ${e}`));
        }
      });

      req.on('error', err => ko(err.message));
    });
  }
}
