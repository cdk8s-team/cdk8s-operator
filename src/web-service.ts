import { Construct } from 'constructs';
import * as awsvpc from './imports/ec2.aws.crossplane.io/vpc';
import * as awssubnet from './imports/ec2.aws.crossplane.io/subnet';
import * as awsrds from './imports/database.aws.crossplane.io/rdsinstance';

export interface CompositePostgreSQLInstanceSpec {
  /**
   * The amount of storage in GB.
   */
  readonly storageGB: number;
}

export class DatabaseComposition extends Construct {
  constructor(scope: Construct, id: string, options: CompositePostgreSQLInstanceSpec) {
    super(scope, id);
    
    const vpc = new awsvpc.Vpc(scope, "my-vpc", {
      spec: {
        forProvider: {
          region: "us-west-2",
          cidrBlock: "192.168.0.0/16",
          enableDnsHostNames: true,
          enableDnsSupport: true,
        }
      }
    });

    const subnets: awssubnet.Subnet[] = [
    new awssubnet.Subnet(scope, "subnet-1", {
      spec: {
        forProvider: {
          region: "us-west-2",
          cidrBlock: "192.168.64.0/18",
          availabilityZone: "us-west-2a",
          vpcIdRef: {
            name: vpc.name,
          },
        }
      }
    }),
    new awssubnet.Subnet(scope, "subnet-2", {
      spec: {
        forProvider: {
          region: "us-west-2",
          cidrBlock: "192.168.128.0/18",
          availabilityZone: "us-west-2b",
          vpcIdRef: {
            name: vpc.name,
          },
        }
      }
    }),
    new awssubnet.Subnet(scope, "subnet-3", {
      spec: {
        forProvider: {
          region: "us-west-2",
          cidrBlock: "192.168.192.0/18",
          availabilityZone: "us-west-2c",
          vpcIdRef: {
            name: vpc.name,
          },
        }
      }
    })]

    const rds = new awsrds.RdsInstance(scope, "my-rds", {
      spec: {
        forProvider: {
          allocatedStorage: options.storageGB,
          region: "us-west-2",
          dbInstanceClass: "db.t2.small",
          engine: "postgres",
          engineVersion: "9.6",
          publiclyAccessible: true,
          skipFinalSnapshotBeforeDeletion: true,
          masterUsername: "admin",
        },
        writeConnectionSecretToRef: {
          name: "my-rds-connection",
          namespace: "crossplane-system",
        }
      }
    })
  }
}