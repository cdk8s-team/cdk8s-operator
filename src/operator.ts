import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

import { App, Chart } from 'cdk8s';
import { Construct } from 'constructs';

/**
 * The handler for this custom resource provider.
 */
export interface ICustomResourceProviderHandler {
  // readonly schema: any;

  apply(scope: Construct, id: string, spec: any): Construct;
}

export interface CustomResourceProvider {
  /**
   * Kind of this custom resource.
   */
  readonly kind: string;

  /**
   * API version of the custom resource.
   *
   * @default "v1"
   */
  readonly apiVersion: string;

  /**
   * The construct handler.
   */
  readonly handler: ICustomResourceProviderHandler;
}

export interface OperatorProps {
  /**
   * A Kubernetes JSON manifest with a single resource that is matched against
   * one of the providers within this operator.
   *
   * @default - first position command-line argument or "/dev/stdin"
   */
  readonly inputFile?: string;

  /**
   * Where to write the synthesized output.
   *
   * @default "/dev/stdout"
   */
  readonly outputFile?: string;
}

/**
 * A CDK8s app which allows implementing Kubernetes operators using CDK8s constructs.
 */
export class Operator extends App {

  private readonly inputFile: string;
  private readonly outputFile?: string;

  private readonly providers: CustomResourceProvider[];

  constructor(props: OperatorProps = {}) {
    const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'cdk8s'));
    super({ outdir });

    this.providers = [];

    this.inputFile = props.inputFile ?? process.argv[2] ?? '/dev/stdin';
    this.outputFile = props.outputFile;
  }

  /**
   * Adds a custom resource provider to this operator.
   * @param provider The provider to add
   */
  public addProvider(provider: CustomResourceProvider) {
    this.providers.push(provider);
  }

  /**
   * Reads a Kubernetes manifest in JSON format from STDIN or the file specified
   * as the first positional command-line argument. This manifest is expected to
   * include a single Kubernetes resource. Then, we match `apiVersion` and
   * `kind` to one of the registered providers and if we do, we invoke
   * `apply()`, passing it the `spec` of the input manifest and a chart as a
   * scope. The chart is then synthesized and the output manifest is written to
   * STDOUT.
   */
  public synth() {
    const input = JSON.parse(fs.readFileSync(this.inputFile, 'utf-8'));

    let write;
    if (this.outputFile) {
      const outfile = this.outputFile;
      write = (data: Buffer) => fs.writeFileSync(outfile, data);
    } else {
      write = (data: Buffer) => process.stdout.write(data);
    }

    if (typeof(input) !== 'object') {
      throw new Error('input must be a single kubernetes resource');
    }

    const provider = this.findProvider(input);

    const name = input.metadata?.name;
    if (!name) {
      throw new Error('"metadata.name" must be defined');
    }

    const namespace = input.metadata?.namespace;

    // TODO: namespace
    const spec = input.spec ?? {};

    const chart = new Chart(this, name, { namespace });

    console.error(`Synthesizing ${input.kind}.${input.apiVersion}`);
    provider.handler.apply(chart, name, spec);

    super.synth();

    for (const file of fs.readdirSync(this.outdir)) {
      const filepath = path.join(this.outdir, file);
      const manifest = fs.readFileSync(filepath);
      write(manifest);
    }
  }

  private findProvider(input: { kind: string; apiVersion: string }) {
    const { apiVersion, kind } = input;

    if (!apiVersion) {
      throw new Error('"apiVersion" is required');
    }

    if (!kind) {
      throw new Error('"kind" is required');
    }

    for (const p of this.providers) {
      if (p.apiVersion === apiVersion && p.kind === kind) {
        return p;
      }
    }

    throw new Error(`No custom resource provider found for ${kind}.${apiVersion}`);
  }
}

