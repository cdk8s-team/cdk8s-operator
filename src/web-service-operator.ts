import { DatabaseComposition } from './web-service';
import { Operator } from './operator'; // <-- this will be in cdk8s

const cr = new Operator();

cr.addProvider({
  kind: 'WebService',
  apiVersion: 'samples.cdk8s.org/v1alpha1',
  // schema: ...
  handler: {
    apply: (scope, name, spec) => new DatabaseComposition(scope, name, spec),
  },
})

cr.synth();