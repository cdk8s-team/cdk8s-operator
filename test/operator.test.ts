import { mkdtempSync, writeFileSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';
import { Yaml } from 'cdk8s';
import { PodCollection } from './fixtures/pod-collection';
import { Operator } from '../src';

// disable logs
jest.spyOn(console, 'error').mockReturnValue();

test('operator with a single provider', () => {
  const workdir = mktmpdir();
  const inputFile = join(workdir, 'input.json');
  const outputFile = join(workdir, 'output.json');

  const operator = new Operator({
    inputFile,
    outputFile,
  });

  operator.addProvider({
    apiVersion: 'foo.bar/v1beta1',
    kind: 'PodCollection',
    handler: {
      apply: (scope, id, spec) => new PodCollection(scope, id, spec),
    },
  });

  // ----------------------------------------------------------------

  writeFileSync(inputFile, JSON.stringify({
    apiVersion: 'foo.bar/v1beta1',
    kind: 'PodCollection',
    metadata: {
      name: 'my-pods',
    },
    spec: {
      image: 'paulbouwer/hello-kubernetes',
      count: 5,
    },
  }));

  operator.synth();

  expect(Yaml.load(outputFile)).toMatchSnapshot();
});

function mktmpdir() {
  return mkdtempSync(join(tmpdir(), 'operator.test.'));
}