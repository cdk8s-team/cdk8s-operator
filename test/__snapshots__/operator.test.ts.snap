// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operator with a single provider 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "name": "my-pods-pod-0-c82b84ba",
    },
    "spec": Object {
      "containers": Array [
        Object {
          "image": "paulbouwer/hello-kubernetes",
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "name": "my-pods-pod-1-c8741f46",
    },
    "spec": Object {
      "containers": Array [
        Object {
          "image": "paulbouwer/hello-kubernetes",
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "name": "my-pods-pod-2-c820d644",
    },
    "spec": Object {
      "containers": Array [
        Object {
          "image": "paulbouwer/hello-kubernetes",
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "name": "my-pods-pod-3-c8c7ccdb",
    },
    "spec": Object {
      "containers": Array [
        Object {
          "image": "paulbouwer/hello-kubernetes",
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "name": "my-pods-pod-4-c81f2857",
    },
    "spec": Object {
      "containers": Array [
        Object {
          "image": "paulbouwer/hello-kubernetes",
        },
      ],
    },
  },
]
`;
